Add support for highlight and linked text annotations

dcaliste opened the pull request:
As discussed in #80 , here is the merge request to add support for highlight and text annotations. I'm using it for days now, and I didn't notice any issues. Maybe, one can start reviewing the C++ code when time is available and leave the UI parts for later, or even for a later PR when improved design will be ready.

What do you think ?

anandrkris commented:
Maybe we  we have an rpm it openrepos or elsewhere for users to test and give feedback?
Also, there is a TJC for this.
https://together.jolla.com/question/25613/feature-request-annotating-and-highlighting-documents/?answer=136640#post-id-136640

dcaliste commented:
Hello, indeed I thought to build an RPM but differ it because:
* I've installed a more recent version of Poppler in my SDK than available on the phone, so build executable won't install properlly on the phone.
* My PR is built on top of master version of sailfish-office which contains some dependencies that are not present on the phone (system icons for documents).

The coming 2.0.2.x version may contain the two prerequisitions mentioned above. Then, I can build an RPM package for everyone to test.

anandrkris commented:
Oh okies. Thanks for clarifying. Appreciate all your good work on this. :smile:

pvuorela commented on pdf/pdflinkarea.cpp:
```diff
@@ -94,18 +156,68 @@ void PDFLinkArea::mouseReleaseEvent(QMouseEvent *event)
+
+    if (d->selection && d->selection->count() > 0) {
+      d->selection->unselect();
+      return;
```
Nitpick: some bad indentation here and in mousePress.

dcaliste replied:
Good catch, emacs wrongly configured ;)

pvuorela commented:
I'll do a limited review while (still) waiting for new UI plan. Also vacation coming up soon.

One thing noted is clicking on highlighted text. Shows press effect of filled rectangle which can be quite much outside the highlight with two column layout. Adobe seems to show only bounding rect edges. Maybe could do something similar.

Related to above, maybe could show only the edges for links too. Now image as link doesn't show as it's supposed to.

pvuorela commented:
Or at least filled rectangle could be more subtle, more transparency.

dcaliste commented:
Thank you for taking time looking at this. I understand that's it's a limited review waiting for design suggestions. Have nice coming holidays, I'll be on vacation also first part of July !

By the way, your remark on visual feedback when tapping on a sensitive part (link or annotation) is exactly the kind of remark that I was waiting for by submitting this early like that. Indeed, you're right, I didn't thought about the two columns spaning annotation but it's really ugly as it is now. Having the visual feedback exactly on each rectangle will imply an important rework for this because as you've seen, currently it's easily implemented with a rectangle property exported to QML.

I'll have to think of something else, but I don't want to through away this visual feedback, I think it brings something additional even for simple links, don't you think ?

dcaliste commented:
Just in case, these commits correspond to work in progress. I still need to polish the highlight context menu and highlight selection from toolbar.

dcaliste commented:
@pvuorela, ok, I think the PR is ready for review when you have time. It implements the following according to design (or what I've understand from design paper, which hopefully corresponds !):
* long press on an empty area open a context menu to create a new text note, opening a dialog.
* an icon in the toolbar is dedicated for text annotations. When tapped, becomes active, keep the toolbar open, display once a helping notification and waits for tap anywhere to create a new note.
* tapping on a note open a page to view it and edit it on tap. It also has a pull down menu to delete the note.
* long tap on a text note opens a context menu to edit or delete the note.
* selecting a text opens the toolbar and an icon can be tapped to highlight the selected text.
* the same icon can be tapped to become active. When active, moving the finger on text select it and when released, the text becomes highlighted. A helping notification is displayed once.
* a short or long tap on an highlighted area opens a context menu to allow to change color or remove the highlight.
* document PDF is updated with modification in place.

In addition to these, I would like to propose some additional features. They are implemented, but they can also be easily removed if desired. I've put them to show case what can be done with these features and hopefully convince that they are usefull and don't clutter the interface too much. It includes:
* displaying the author of a text annotation as title of the text annotation page, leaving it empty as design recommends when there is no author.
* display as two lines maximum the text content (*i.e.* comment) of an highlight, when any, at the top of the highlight context menu. So one can easily have a clue of the highlight reason by tapping on the highlight. Tapping on this menu item opens the annotation page so one can view the full text and edit it if necessary. When there is no text content for an highlight, I've added a menu item that proposes to add a comment on  the highlight, opening the annotation dialog.
* I've added the highlight style possibilities (*e.g.* underlines, strike through…) in the highlight context menu under the color choices.

I've taken your first remark into account for visual feedback on large annotations. Now, just the tapped rectangle inside an annotation is highlighted when tapped.

Finally, I've one or two questions: what do you recommend concerning the size of PDFDocumentPage.qml file ? The toolbar and the various context menus take quite a big number of lines. Would you recommend to move them in dedicated qml files, or is it just fine line that ?

pvuorela commented:
Starting to look quite nice with the new design!

...But did find a bunch of little details which work funny or we could adjust the behavior :)

- Press & hold should allow to move finger down and select menu items
- Selecting "Add note" from menu flashes "Delete" while dialog is transitioning visible
- Many times got menu into state where it thinks I'm editing existing annotation even though where I press&hold there is none
- Deletion remorse timer can be removed
- Note editing dialog should have editor size until keyboard top even if empty -> clicking near "Write a note" should focus the editor.
- Opening an existing note looks actually quite empty. Talked with Martin that we could perhaps have then some page header like "Note".
- Clicking on new note button in toolbar should clear selection. Now it leads into strange state when there's at the same time selection and tap creates a note.
- Need new icons for note button, but I can handle that later.
- If user changes color of a highlight, we could store the color in dconf to be user for new highlights coming after that. If user prefers something else than yellow, doesn't need to keep selecting it.
- Maybe same for underlining style if that's kept
- Could maybe mark selected underlining style with highlight color?
- Create annotation on first page and try dragging down -> page content dims except lighter rectangle around the annotation.
- Ideally highlight should continue over words in a row without bumps/spaces between words. Not sure how well Poppler supports, but at least Evince does that. Roundings on highlight edges are not specified but can be kept. For one detail the press effect is smaller than word's highlight.
- Zoom near annotation and click it -> black rectangle appears instead of context menu.
- Comment in context menu not entirely according to UI conventions. If it's clickable it should be primary color. Also ponder if it needs some header/icon to indicate what it is. Might need some thinking more.

Size of file doesn't matter to me that much.

dcaliste commented:
Great, thank you for all the interesting notes. I'll address the simple minor (but relevent) details tomorrow or during the week-end and report here when done. But before doing so, some remarks about some of your concerns:
> Press & hold should allow to move finger down and select menu items

I agree, and have noticed this also, but didn't find any solution. Ok, I'll try harder ! Something to do with mouse.accept in the press signal handler of pdflinkarea maybe ?

> Many times got menu into state where it thinks I'm editing existing annotation even though where I press&hold there is none

The wiggle for detection is quite large (Theme.itemSizeMedium / 2), do you notice this behaviour within that range, or far outside ?

> Opening an existing note looks actually quite empty. Talked with Martin that we could perhaps have then some page header like "Note".

I agree, "note" or something if no author, or the author.

> Create annotation on first page and try dragging down -> page content dims except lighter rectangle around the annotation.

The most interesting bug. I've noticed it also, but only when rotating the device after having done an annotation. During rotation, the buggy state is not kept for long so I didn't understand the origin of it. But with your way to trigger it, I've now understood. The annotation, for quick rendering, is done by patches on the big page texture. Patches are thrown away when the page is rerendered. In the patch state, the page is thus the sum of several GSSimpleTextureNode. When the screen is dimed, each textures are rendered with an opacity < 1, thus patches are more visible, because they are the sum of the page behind and themselves. I have no idea how to simply solve this… Random thoughts giving: making holes in the page texture where the patches are, asking to rerender the full page after the patch. But none of these is satisfactory.

> Ideally highlight should continue over words in a row without bumps/spaces between words. Not sure how well Poppler supports, but at least Evince does that.

That's the old selection issue we have already discussed together. Currently, the highlight annotation implementation I've done, relies on the selection rectangles to define the quads used by Poppler for the highlight. But the selection rectangles are currently on words only due to Poppler-Qt limitations. There are routines in Poppler-Glib to obtain continuous selections between words, but we're not using them (not available from Poppler-Qt). The idea is to later on, import the code of Poppler-Glib which is doing plain Poppler to get continuous selection and use this instead of the selection by word of Poppler-Qt. After having continuous selection, the highlight will automatically works out of the box. Ideally, I would like to upstream this to Poppler-Qt, but that's another story.

> If it's clickable it should be primary color. Also ponder if it needs some header/icon to indicate what it is. Might need some thinking more.

The other solution, would be to keep the "add/edit a comment" menu item all the time (with a different label). Like that, we keep the small summary on top of the menu, close to the highlight (a bit like the URL when cliking on links), giving a quick and unobstrusive view on the reason of the highlight, if any.

dcaliste commented:
> Zoom near annotation and click it -> black rectangle appears instead of context menu.

This is the case for links also. It's due to ContextMenuHook, so I've created a specific pull request to correct this independently from the annotation PR, see #90.

pvuorela commented:
> > Press & hold should allow to move finger down and select menu items

> I agree, and have noticed this also, but didn't find any solution. Ok, I'll try harder ! Something to do with mouse.accept in the press signal handler of pdflinkarea maybe ?

Context Menu should support when inside a MouseArea. Though it might get tricky here when menu parent is emulated with the hook thing. Won't hold back merging because of this.

> > Many times got menu into state where it thinks I'm editing existing annotation even though where I press&hold there is none

> The wiggle for detection is quite large (Theme.itemSizeMedium / 2), do you notice this behaviour within that range, or far outside ?

Far outside. E.g. press and hold on top of first page to create note, after creating press and hold on the next page.

> >  If it's clickable it should be primary color. Also ponder if it needs some header/icon to indicate what it is. Might need some thinking more.

> The other solution, would be to keep the "add/edit a comment" menu item all the time (with a different label). Like that, we keep the small summary on top of the menu, close to the highlight (a bit like the URL when cliking on links), giving a quick and unobstrusive view on the reason of the highlight, if any.

Yea, option too.

dcaliste commented:
I've addressed the following remarks:
> Press & hold should allow to move finger down and select menu items

I've added a special commit for this. For it to work, as you said, the PDF.LinkArea object needs to emit the positionChanged and released signals. Then, they are catched by the ContextMenuHook and reemit there for the ContextMenu to work. Special notice: it's not working when the background item is not occupying the full width (like the highlight style) because ContextMenu.qml detect the item only with its y coordinate.

> Selecting "Add note" from menu flashes "Delete" while dialog is transitioning visible

Corrected.

> Deletion remorse timer can be removed

Removed.

> Note editing dialog should have editor size until keyboard top even if empty -> clicking near "Write a note" should focus the editor.

Done like in the note application.

> Opening an existing note looks actually quite empty. Talked with Martin that we could perhaps have then some page header like "Note".

Done like that. "Note" is displayed if no author or the author name if exists.

> Clicking on new note button in toolbar should clear selection. Now it leads into strange state when there's at the same time selection and tap creates a note.

Now any action on toolbar clear the selection.

> If user changes color of a highlight, we could store the color in dconf to be user for new highlights coming after that. If user prefers something else than yellow, doesn't need to keep selecting it.

Added a dconf entry and reusing it.

> Maybe same for underlining style if that's kept

Same for style.

> Could maybe mark selected underlining style with highlight color?

Corrected.

> Comment in context menu not entirely according to UI conventions. If it's clickable it should be primary color. Also ponder if it needs some header/icon to indicate what it is. Might need some thinking more.

Corrected by enabling the add / edit menu entry all the time and removing the background item under the info label.

This item is still pending:

> Many times got menu into state where it thinks I'm editing existing annotation even though where I press&hold there is none

Sorry, even with your comment before, I cannot reproduce this behaviour. Is it only on specific PDF documents or on all ?

I think I've otherwise addressed all the remarks you've done. I will continue to test this to check the various git-rebase didn't introduce any regressions and see if there are some unseen corner cases.

pvuorela commented on pdf/pdfannotation.cpp:
```diff
@@ -0,0 +1,353 @@
+    if (icon() == value)
+        return;
+
+    int index = (int)value;
```
Cosmetics but generally c++ style casts preferred. static_cast<int>(value). Few other instances elsewhere too.

pvuorela commented on pdf/pdfannotation.cpp:
```diff
@@ -0,0 +1,353 @@
+
+    int index = (int)value;
+    if (index > iconNames.length() || index < 0) {
+        qWarning() << QString::fromLocal8Bit("Wrong icon for text annotation (%1).").arg(index);
```
QStringLiteral could be better here and elsewhere.

dcaliste replied:
That's done here and in pdfrenderthraed.cpp also.

pvuorela commented on pdf/pdfannotation.h:
```diff
@@ -0,0 +1,189 @@
+
+private:
+    class Private;
+    Private *d;
```
No change needs, but just to note that private data classes can be just omitted since it doesn't matter on qml interfaces nor inside one application.

dcaliste replied:
Ok, I take note. I think you understand now that I'm learning C++ with this project ;)

pvuorela replied:
Sure :) The reason why it's usually used is keeping binary compatibility, added/removed member variables change size of a class instance. But since this is not a library external software links against, it doesn't provide much.

dcaliste replied:
Thanks for the explanation. It makes sense.

pvuorela commented on pdf/pdfdocument.cpp:
```diff
@@ -65,6 +68,11 @@ QString PDFDocument::source() const
     return d->source;
 }
 
+QString PDFDocument::autoSave() const
```
Suggestion: autoSaveName wouldn't sound as an action.

dcaliste replied:
I've changed it for ```autoSavePath```, because it receives something starting with file:///.

pvuorela commented on plugin/PDFAnnotationEdit.qml:
```diff
@@ -0,0 +1,71 @@
+                placeholderText: qsTrId("sailfish-office-ta-text-annotation")
+                onTextChanged: {
+                    if (annotation) {
+                        annotation.contents = text                
```
Cosmetics: trailing spaces :)

dcaliste replied:
Removed.

dcaliste commented:
I've rebased all commits on master.

pvuorela commented on plugin/PDFView.qml:
```diff
@@ -214,17 +220,32 @@ SilicaFlickable {
+                selection: pdfSelection
+
+                onPositionChanged: {
+                    console.log(at)
```
Looks like a leftover.

dcaliste replied:
Indeed !

pvuorela commented on plugin/PDFSelectionView.qml:
```diff
@@ -23,6 +23,7 @@ Repeater {
     id: root
 
     property Item flickable
+    property bool dragable: true
```
draggable

dcaliste replied:
Oups, sorry, should have paid better attention to English at school…

pvuorela commented on plugin/PDFAnnotationEdit.qml:
```diff
@@ -0,0 +1,71 @@
+                background: null
+                focus: false
+                text: annotation ? annotation.contents : ""
+                //% "Write a note…"
```
For highlight this and header should probably say "comment", action is "Add a comment".

dcaliste replied:
That's logical. I've changed the title and the placeholder to be refering to "note" when dealing with Text or Caret annotations and refering to "comment" for any other annotations.

pvuorela commented on plugin/PDFDocumentPage.qml:
```diff
@@ -108,30 +109,105 @@ DocumentPage {
+            hook.showMenu(contextMenuText)
+        }
+        clip: anchors.bottomMargin > 0 || base.status !== PageStatus.Active
+        RemorsePopup {
```
Looks unused.

dcaliste replied:
That's true, another leftover, so removed.

pvuorela commented on plugin/PDFDocumentPage.qml:
```diff
@@ -189,9 +375,182 @@ DocumentPage {
+                id: colors
+                model: ["#db431c", "#ffff00", "#8afa72", "#00ffff",
+                "#3828f9", "#a328c7", "#ffffff", "#989898",
+                "#000000"]
```
Cosmetics, but line splitting with same indentation as property declaration looks strange. Ideally would match with first item in list.

dcaliste replied:
Corrected. Normally, I do like that, but maybe after several code moves Emacs decided differently and I didn't correct it afterwards.

pvuorela commented on plugin/PDFDocumentPage.qml:
```diff
@@ -108,30 +109,105 @@ DocumentPage {
+                case PDF.Annotation.Caret:
+                case PDF.Annotation.Text:
+                contextMenuText.annotation = annotation
+                hook.showMenu(contextMenuText)
```
Cases and actual code shouldn't start on same point. Qt convention would be moving case to match with switch.

dcaliste replied:
Realigned at the three places in PDFDocumentPage.qml.

pvuorela commented on pdf/sailfishofficepdfplugin.cpp:
```diff
@@ -35,4 +36,8 @@ void SailfishOfficePDFPlugin::registerTypes(const char *uri)
+    qmlRegisterType<PDFAnnotation>(uri, 1, 1, "Annotation");
+    qmlRegisterType<PDFTextAnnotation>(uri, 1, 1, "TextAnnotation");
+    qmlRegisterType<PDFCaretAnnotation>(uri, 1, 1, "CaretAnnotation");
+    qmlRegisterType<PDFHighlightAnnotation>(uri, 1, 1, "HighlightAnnotation");
```
I think I'd just keep 1.0 for simplicity. Shouldn't be imported by anyone except this app. But doesn't really matter.

dcaliste replied:
Ok, I think you're right. Since I was adding new QML object in the library, I increased the version number, but SailfishOffice being the only user of this library, it makes no sense to bother with the versioning. Switched back to 1.0.

pvuorela commented on plugin/PDFDocumentPage.qml:
```diff
@@ -189,9 +375,182 @@ DocumentPage {
+        }
+    }
+
+    ContextMenu {
```
Can do later, but having this created dynamically when needed would save a bit memory and cpu time on the common case when annotations are not used. Additionally in separate file would save even more as less qml parsing would be needed then.

pvuorela commented on plugin/PDFDocumentPage.qml:
```diff
@@ -189,9 +375,182 @@ DocumentPage {
+                model: [{"style": PDF.HighlightAnnotation.Highlight,
+                         "label": "abc"},
+                        {"style": PDF.HighlightAnnotation.Squiggly,
+                         "label": "a&#816;b&#816;c&#816;"},
```
Hmh, wonder if &#x330; would be easier to check what the character is, unicode more commonly presented in hex.

pvuorela replied:
Ehm, was supposed to be 816 -> x330

pvuorela commented on pdf/pdflinkarea.cpp:
```diff
@@ -148,8 +192,50 @@ void PDFLinkArea::mouseReleaseEvent(QMouseEvent *event)
+    } else if (d->selection && d->selection->count() > 0) {
+        d->selection->unselect();
+    } else if (d->annotation != nullptr) {
+        // proxy will be child of this object to avoid memory leak.
```
Means also that annotation proxies will be accumulated as long as the document is open, right? Wonder would this work better if ownership is passed to qml engine (QQmlEngine::setObjectOwnership() etc). Didn't test though.

pvuorela replied:
May or may not work, but some "activeAnnotation" property might be an option too.

dcaliste replied:
That's right. At the time I wrote this, I read extensively the Qt documentation and I remember reading that transfer ownership to QML was impossible: http://doc.qt.io/qt-5/qtqml-cppintegration-data.html <quote>When data is transferred from C++ to QML, the ownership of the data always remains with C++.</quote>

So yes, all the annnotation wrappers, generated when tapping on an annotation, will stay alive for the duration of the document. I don't think that's really a memory penalty, even on small memory systems. Indeed, the wrapper is only sizeof(QObject) + some pointers and one int and one boolean. So we're speaking of some kilobytes at most even if the user is tapping on dozen of annotations. Don't you think so ?

pvuorela replied:
Yea, not big objects so could perhaps tolerate, but should have some remark about that at least.

dcaliste replied:
I've added the following comment:
>  Todo: transfer ownership to QML, if possible. Before that,
     all created proxy objects will be alive up to the moment
     the document object is released.

pvuorela commented on pdf/pdfannotation.cpp:
```diff
@@ -0,0 +1,353 @@
+
+PDFAnnotation::SubType PDFAnnotation::type() const
+{
+    return (PDFAnnotation::SubType)d->m_annotation->subType();
```
Cast style.

pvuorela replied:
Also icon() and style() later.

pvuorela commented on pdf/pdfannotation.cpp:
```diff
@@ -0,0 +1,353 @@
+{
+    attachOnce(document, page);
+
+    d->m_annotation->setBoundary(QRectF(x, y, 24., 24.));
```
Magic value.

dcaliste replied:
Yes, I've explained this in one of the issue #80 comments. It is the size Poppler gives to the text annotation icon that is added in the document. It's also hard-coded in Poppler, and I didn't find any mean to retrieve this value.

I'll add a comment in our code about this.

pvuorela commented on pdf/pdfannotation.h:
```diff
@@ -0,0 +1,189 @@
+    Q_PROPERTY(QDateTime modificationDate READ modificationDate NOTIFY modificationDateChanged)
+    Q_PROPERTY(QColor color READ color WRITE setColor NOTIFY colorChanged)
+    Q_PROPERTY(SubType type READ type CONSTANT)
+
```
Some unused properties here. Could just remove? Also iconType on text annotation.

dcaliste replied:
You're right. Initially, I've bound as much of Poppler-Qt Annotation object as possible from the C++ side, because it was 1:1 code. Then, the UI part is finally not using all capabilities exposed from C++. I didn't have envy to just erase all this code, even if not use, in case it will be later.

From code development point of view, it's dead code, indeed. I will remove it if you think it's better like that (*i.e.* less maintenance burden).

pvuorela commented on plugin/PDFDocumentPage.qml:
```diff
@@ -108,30 +109,102 @@ DocumentPage {
+        }
+
+        Notification {
+            id: notice
```
Also a potential candidate for lazy creation.

dcaliste replied:
That's true.

pvuorela commented on plugin/PDFDocumentPage.qml:
```diff
@@ -144,6 +217,113 @@ DocumentPage {
+                    icon.source: "image://theme/icon-m-edit"
+                    onClicked: highlightTool.clicked(mouse)
+                }
+                Component {
```
Suggest moving this as first or last child element, now between the directly instantiated ones.

dcaliste replied:
Ok, moved it first because it's called from the function that is on top of this BackgroundItem.

pvuorela commented on pdf/pdfselection.cpp:
```diff
@@ -570,6 +572,34 @@ void PDFSelection::onLayoutChanged()
+{
+    if (d->pageIndexStart < 0 || d->pageIndexStop < 0 ||
+        d->boxIndexStart < 0 || d->boxIndexStop < 0)
+        return false;
```
Should have curly braces if conditions wrap multiple lines. || at start of line.

dcaliste commented:
Ok, hopefully, I should have addressed all your remarks (thanks a lot for your full afternoon review !), except these ones:
* annotation wrappers piling up on C++ side for the full life duration of the document object (I can test the currentAnnotation property solution if you think it's worth).
* unused properties in pdfannotation.h (wait for your decision to remove them or not).
* add a commit to dynamically create the contextmenus and notifications (need to further work on this).

dcaliste commented:
As a side note, the unused properties concern this (for you to have a better view of what could be valuable in futur evolutions):
* in PDFAnnotation, creation and modification dates, may be displayed in the PDFAnnotationEdit.qml page;
* in PDFTextAnnotation, the icon, may be used to display a different icon instead of the note one (can be a question mark, a help icon…).

If I'm not doing any mistake, all other properties are currently used here and there.

pvuorela commented on plugin/PDFAnnotationNew.qml:
```diff
@@ -0,0 +1,53 @@
+                width: parent.width
+                height: Math.max(flickable.height - dialogHeader.height, implicitHeight)
+                //% "Write a note…"
+                placeholderText: qsTrId("sailfish-office-ta-text-annotation")
```
Comment text version here too.

dcaliste replied:
That's right, I thought this one was used only for text annotations. It's corrected.

pvuorela commented on plugin/PDFView.qml:
```diff
@@ -214,17 +220,29 @@ SilicaFlickable {
+                onAnnotationLongPress: base.annotationLongPress(annotation, contextHook)
+                onTextAnnotationLongPress: base.annotationLongPress(annotation, contextHook)
+                onCaretAnnotationLongPress: base.annotationLongPress(annotation, contextHook)
+                onHighlightAnnotationLongPress: base.annotationLongPress(annotation, contextHook)
```
Why different signals for different types of annotations if they all end up handled the same and annotation itself can also identify the type?

dcaliste replied:
I've tested the following, but it didn't work. Maybe you can spot a mistake somewhere.
pdflinkarea.h:
signal annotationPress(PDFAnnotation *anno);

pdflinkarea.cpp:
        switch (d->annotation->subType()) {
        case Poppler::Annotation::SubType::AText: {
            PDFTextAnnotation *proxy = new PDFTextAnnotation(…);
            emit annotationPress(static_cast<PDFAnnotation*>(proxy));
            return;
        }
…

PDFDocumentPage.qml:
onAnnotationPress: if (anno.type == PDF.Annotation.Text) console.log(anno.icon)

This raises a runtime error, saying that icon is not a property of type PDFAnnotation. So what I've deduced is that, having a signal with signature (Annotation *anno), make the QML part loose the information that this anno object is indeed a specialized text version of annotation. Thus, I decided to create as many signals with detailled signature as I have of specialized types, then I store all these different QML object into a single variant QML property, so the type is preserved even if accessed from the same name.

pvuorela replied:
Changed myself textAnnotationClicked to annotationClicked and works for me, click handler outputs icon type. Casting unnecessary, but don't think it should matter.

dcaliste replied:
Ok, I will retry after lunch to use the unified signal emission from pdflinkarea and report if I stumble on the same issue as before.

dcaliste replied:
Indeed, it's working. So I've removed all the differentiated signals from pdflinkarea.h. I don't remember what specific code implementation I did that triggered the above issue.

pvuorela commented on pdf/pdfannotation.cpp:
```diff
@@ -0,0 +1,356 @@
+    d->m_document->removeAnnotation(d->m_annotation, d->m_page);
+    d->m_annotation = nullptr;
+
+    deleteLater();
```
Since invokable, could consider checking m_annotation null pointer and always deleting.

pvuorela commented on pdf/pdfannotation.cpp:
```diff
@@ -0,0 +1,356 @@
+void PDFAnnotation::attach(PDFDocument *document, PDFSelection *selection)
+{
+    if (!selection)
+        return;
```
For robustness: || selection->count() == 0?

dcaliste replied:
Very true, added.

pvuorela commented on pdf/pdfdocument.cpp:
```diff
@@ -144,6 +157,48 @@ void PDFDocument::setSource(const QString &source)
+    documentModified();
+}
+
+void PDFDocument::documentModified()
```
Suggestion: setDocumentModified(), modified alone sounds like a signal.

pvuorela commented on pdf/pdfannotation.cpp:
```diff
@@ -0,0 +1,356 @@
+    emit colorChanged();
+
+    if (d->m_document != nullptr && d->m_page >= 0)
+        d->m_document->onPageChanged(d->m_page, d->m_annotation->boundary());
```
Cannot modify page without touching the document it belongs to, right? Could then make this method trigger modification handling too and remove all documentModified() calls that are done afterwards.

dcaliste replied:
Good reasoning. I've simplified the implementation by calling the setDocumentModified() in the onPageModified() slot, so setDocumentModified() is not needed anymore in pdfrenderthread when adding or removing an annotation and in the three places where modification on annotation modify the page in pdfannotation.cpp.

pvuorela commented on pdf/pdfcanvas.cpp:
```diff
@@ -350,6 +376,57 @@ QPair<QUrl, PDFCanvas::ReducedBox> PDFCanvas::urlAtPoint(const QPointF &point) c
+QPair<Poppler::Annotation*, PDFCanvas::ReducedBox> PDFCanvas::annotationAtPoint(const QPointF &point) const
+{
+    for (int i = 0; i < d->pageCount; ++i) {
+        const PDFPage &page = d->pages.value(i);
```
Guess in reality doesn't matter that much, but a bit concerned that now on clicks the document pages are traversed for both urls and links. Can do later, but should consider how to make page lookup and files with ridiculous amount of pages in general more efficient.

dcaliste replied:
Yes, I guess a dichotomic algorithm should switch from O(N) to O(ln(N)) since the rectangles of pages are sorted (and almost all of the same size). I mean, int(at.y / height) should give a good guess of page index i0, then trying i0 + int((at.y - i0.y)/(height-i0.height)…

pvuorela commented on pdf/pdfdocument.cpp:
```diff
@@ -162,6 +217,16 @@ void PDFDocument::requestPage(int index, int size, QQuickWindow *window, QRect s
+        return;
+
+    RenderPageJob* job = new RenderPageJob(index, size, window, subpart,
+                                           RenderPageJob::RenderPageType::Area);
```
This seems a bit silly that there are separate methods for requesting arbitrary areas of pages that are considered root texture and arbitrary areas that are considered patches. Could be nice if patch approach would be implementation detail of pdfcanvas. E.g. add some int extraData to requestPage and return that on finish signal.

dcaliste replied:
I did it like that to put more sense into the API, even if a bit more clumpsy. In fact the rendering job is already unified with an enum to decide between Page and Area. The best would have been to expose this enum in pdfcanvas which does the request and pdfjob that stores it, but pdfjob is not visible from pdfcanvas which only address pdfdocument.

What you suggest would have the following API (correct me if I misinterpret your suggestion):
pdfcanvas.cpp:
enum {Page, Area};
m_document->requestPage(…, Page) or m_document->requestPage(…, Area)

pdfdocument.h:
requestPage(…, int flag);

pdfjob.h:
RenderPageJob(…, int flag);

IMHO, when reading this API, the pro is that its simple, to request a page, call requestPage() and details are implementation specific hidden in flag argument. The con is that flag has no mean when not explained from pdfcanvas. What do you prefer ?

dcaliste replied:
Ok, let's go for the ```, gpointer user_data)``` that I know well ! Joke aside, I suppose in your response that you mean « pdfjob doesn't need to know meaning if it only provides a method for storing random data », instead of « pdfcanvas…»

I'll switch for an API like suggested before this evening, thank you for your answer.

dcaliste replied:
Ok, done.

dcaliste commented:
While reviewing your remarks, I noticed that I introduced a regression on the last commit, *i.e.* the tap for long press and move is not working anymore and code to allow this disappear in rebase. I'll try to recover from home backup this evening. So currently don't pay attention to this not working anymore. Sorry for the inconvenience.

pvuorela commented:
Hmh, now press&hold and then move finger on top of menu doesn't allow to select "add note". Think it worked yesterday at some point? On existing annotation there seems to be some offset. Was earlier seeing also that moving finger on top of color or highlight type kept below item selected.

The case of creating patch on first page and dragging against document start remains. But yeah, don't have good ideas at the moment what to do. Could at least document the shortcoming as some fixme.

pvuorela commented on pdf/pdfdocument.cpp:
```diff
@@ -37,7 +37,9 @@ class PDFDocument::Private
     QString source;
+    QString autoSavePath;
     bool completed;
+    bool modified;
```
Missing initialization?

dcaliste replied:
true, corrected.

pvuorela commented on pdf/pdflinkarea.cpp:
```diff
@@ -30,14 +31,18 @@ class PDFLinkArea::Private
+
     QUrl link;
+    Poppler::Annotation *annotation;
+    bool clickOnSelection;
```
Better initialize these too on Private class ctor.

dcaliste replied:
Ok, the selection attribute was missing initialisation also.

pvuorela commented on pdf/pdflinkarea.cpp:
```diff
@@ -148,8 +196,53 @@ void PDFLinkArea::mouseReleaseEvent(QMouseEvent *event)
+            PDFAnnotation *proxy = new PDFAnnotation
+                (d->annotation, d->canvas->document(), d->pressedBox.first, this);
+            emit annotationClicked(proxy);
+            return;
```
Just a minor thing, but all cases emit same signal, can as well do that outside the switch.

dcaliste commented on pdf/pdfannotation.cpp:
```diff
@@ -0,0 +1,346 @@
+
+void PDFAnnotation::remove()
+{
+    if (d->m_document  && d->m_annotation) {
```
Oops, deleted.

pvuorela commented on pdf/pdfannotation.h:
```diff
@@ -0,0 +1,189 @@
+
+    friend class PDFTextAnnotation;
+    friend class PDFCaretAnnotation;
+    friend class PDFHighlightAnnotation;
```
Also option to use protected instead of private.

dcaliste replied:
Change from friend to protected to allow inheritage.

pvuorela commented on pdf/pdflinkarea.cpp:
```diff
@@ -132,6 +180,36 @@ void PDFLinkArea::mouseMoveEvent(QMouseEvent *event)
 
+PDFAnnotation* PDFLinkArea::newProxyForAnnotation()
+{
+    PDFAnnotation *proxy;
```
Unused.

pvuorela commented on pdf/pdfselection.cpp:
```diff
@@ -570,6 +572,35 @@ void PDFSelection::onLayoutChanged()
+    if (!doc)
+        return false;
+    
+    int i;
```
No need to have outside for loop?

dcaliste commented:
Ok, back at home, I found my backup and restored the commit allowing to continuously move the finger to select item in the context menu.

Besides, I've also found the root of the issue you reported that long press anywhere seems to believe that an annotation was pressed. To reproduce:
* first tap on a note and discard the edit page ;
* long tap anywhere, the context menu was kept showing at the right place but for the previously tapped note.
It's because the annotation property of the context menu was not reseted.

I'll do the protected declarations and the change in requestPage API later this evening.

dcaliste commented:
I've added an additional commit where the notifications and the context menus are loaded dynamically on demand. Context menus have also been moved to separated files.

dcaliste commented:
Last remark: I noticed also that the highlight of the menu items in the highlight context menu is not working well when moving the finger above the color line or the style line. Indeed, it comes from the code in Silica/ContextMenu.qml. There, the highlight logic is done by looking for a child of the context menu at a given y position *and* that has a "__silica_menu_item" property. There is an asumption there that a child menu item should always be found, so the there is no highlight removing in case none is found.

I can cheat a bit by adding a dummy "__silica_menu_item" property on the color line and style line. But even like that, for the style line, the code in ContextMenu.qml is missing the x information, so we cannot select a given style inside the line.

pvuorela commented on plugin/PDFContextMenuHighlight.qml:
```diff
@@ -0,0 +1,137 @@
+                    textFormat: Text.RichText
+                    color: bgStyle.highlighted
+                           || (contextMenuHighlight.annotation !== undefined
+                               && contextMenuHighlight.aannotation !== null
```
Now that you mentioned these, property variant is deprecated and should be changed to property var: https://wiki.qt.io/Property-var - besides annotations, there's property variant _menu on ContextMenuHook.

For import, don't think there's anything requiring "as PDF", so just drop it if you want.

pvuorela commented on plugin/PDFContextMenuHighlight.qml:
```diff
@@ -0,0 +1,137 @@
+        id: infoContents
+        visible: infoContents.text != ""
+        width: parent.width
+        anchors.topMargin: Theme.paddingSmall
```
Not top anchored, so don't think this does anything. And cannot be vertically anchored anyway since ContextMenu puts this inside a Column. verticalAlignment: Text.AlignVCenter could do what's intended.

pvuorela commented on plugin/PDFDocumentPage.qml:
```diff
@@ -105,33 +109,138 @@ DocumentPage {
 
+        function noticeHide() {
+            if (!notice) {
+                notice = noticeComponent.createObject(toolbar)
```
If no notification existed, there's nothing to hide?

pvuorela commented:
> I can cheat a bit by adding a dummy "__silica_menu_item" property on the color line and style line. But even like that, for the style line, the code in ContextMenu.qml is missing the x information, so we cannot select a given style inside the line.

Yea, not necessarily important to be able to select color or style by dragging, highlight not very visible anyway with such small items, fitting better for full width ones. Not selecting wrong items would be good enough, but even that is quite minor.

pvuorela commented:
I think this is starting to be ready, but were you still planning on adjusting those "as PDF" / property variant -> var things?

dcaliste commented:
> I think this is starting to be ready

Great ! I understand that I still have a long way to go before proposing PR without so many tiny errors or mistakes. Thank you again for the long and deep review.

> were you still planning on adjusting those "as PDF" / property variant -> var things?

Sorry, I missed your reply to the question above, so I was waiting. It's updated now. I've switch from variant to var in ContextMenuHook and I've used property Annotation everywhere in the PDFContextMenu*.qml, simplifying the tests also.

pvuorela commented:
> Great ! I understand that I still have a long way to go before proposing PR without so many tiny errors or mistakes. Thank you again for the long and deep review.

Also difficult to even avoid review changes with this big bunch of code.

> Sorry, I missed your reply to the question above, so I was waiting. It's updated now. I've switch from variant to var in ContextMenuHook and I've used property Annotation everywhere in the PDFContextMenu*.qml, simplifying the tests also.

Hmh, looks still variants to me?

dcaliste commented:
Arg, made a second time the mistake of rebasing the wrong tree when switching between machines… Corrected now.

pvuorela commented:
Still found another issue: I have big pdf, 238 pages, highlight annotation on the second page. Changed the color, quickly popped the page, reopened the document -> old color shown. If then reopen again, the color is changed. Suppose writing the new pdf file is still unfinished when I open the document fast. Can also see smaller file sizes if I do change, pop, reopen, pop, etc.

But I think that's corner case enough not to block merging. Great feature here, huge thanks for implementation!
